Event Timing: implement interactionId logic

francovs opened the pull request:
#### 5e3ae48082f5d4d00eb06990803e62cea40ce0e1
<pre>
Event Timing: implement interactionId logic
<a href="https://bugs.webkit.org/show_bug.cgi?id=297361">https://bugs.webkit.org/show_bug.cgi?id=297361</a>
<a href="https://rdar.apple.com/158264903">rdar://158264903</a>

Reviewed by Ryosuke Niwa.

Assigns interactionId to PerformanceEventTiming entries that qualify,
according to <a href="https://www.w3.org/TR/event-timing/.">https://www.w3.org/TR/event-timing/.</a>

Only certain deliberate user actions such as clicks or keypresses are
considered; passive events such as pointerover or pointerexit do not
qualify. Some events are grouped together, receiving the same
interactionId. One such example is pointerdown, pointerup and click.

We currently only support a single pointer, so the map variables
related to pointer interactions (pointerMap[] and pendingPointerDown[])
mentioned in the spec (<a href="https://www.w3.org/TR/event-timing/#pointer-interaction-value-map)">https://www.w3.org/TR/event-timing/#pointer-interaction-value-map)</a>
were implemented as a struct (PointerInteractionState) containing
the necessary state for a single pointer. In that conversion,
pointerMap was renamed to PointerInteractionState::interactionID.

New files created to achieve more fine grained header inclusion:
* Source/WebCore/page/EventTimingInteractionID.h
* Source/WebCore/page/PerformanceEventTimingCandidate.h

Canonical link: <a href="https://commits.webkit.org/299033@main">https://commits.webkit.org/299033@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/787fe6aeaef6742305412c5176f0b7d8d42edfbf

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/117541 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/37217 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/27841 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/123652 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/69548 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/119419 "Passed tests") | [  ~~🛠 ios-sim~~](https://ews-build.webkit.org/#/builders/130/builds/37911 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/45801 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/89208 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/69548 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/120493 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/130/builds/37911 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/105395 "Passed tests") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/41/builds/69713 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/130/builds/37911 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/23512 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/67318 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/130/builds/37911 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/136/builds/23694 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [❌ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/126766 "Failed to checkout and rebase branch from PR 49304") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/44444 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/33438 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [❌ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/2/builds/126766 "Failed to checkout and rebase branch from PR 49304") | 
| | [  ~~🛠 vision-sim~~](https://ews-build.webkit.org/#/builders/121/builds/44802 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [❌ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/101625 "Exiting early after 10 failures. 11 tests run. 1 failures") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/2/builds/126766 "Failed to checkout and rebase branch from PR 49304") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/24861 "Built successfully and passed tests") | [⏳ 🧪 vision-wk2 ](https://ews-build.webkit.org/#/builders/visionOS-2-Simulator-WK2-Tests-EWS "Waiting in queue, processing has not started yet") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/137/builds/20981 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/40802 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/127/builds/44315 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/49990 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/125/builds/43772 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/129/builds/47121 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/124/builds/45465 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

rreno commented on Source/WebCore/page/LocalDOMWindow.h:
```diff
@@ -440,6 +442,17 @@ class LocalDOMWindow final
+    std::optional<PerformanceEventTiming::InteractionID> m_pointerInteractionId;
+    std::optional<PerformanceEventTiming::Candidate> m_pendingPointerDown;
+
+    WTF::StdUnorderedMap<int, PerformanceEventTiming::Candidate> m_pendingKeyDowns;
```
Is there a reason you used this container instead of a HashMap?

francovs replied:
Not a good reason... changing to `HashMap`.

I wanted some map wrapper that simply does a linear seek on top of a `Vector<pair<int, Candidate>>`, since this map should contain very few short-lived entries. I now realize `StdUnorderedMap` is not that...

rreno commented on Source/WebCore/page/LocalDOMWindow.h:
```diff
@@ -440,6 +442,17 @@ class LocalDOMWindow final
+    WTF::StdUnorderedMap<int, PerformanceEventTiming::Candidate> m_pendingKeyDowns;
+    PerformanceEventTiming::InteractionID m_userInteractionValue = 100;
+    unsigned m_interactionCount = 0;
+    void increaseInteractionCount();
```
Please don't mix member variables with function declarations. Generally the private function declarations should be before member variable declarations.

rreno commented on Source/WebCore/page/Performance.cpp:
```diff
@@ -293,25 +311,25 @@ void Performance::countEvent(EventType type)
-void Performance::processEventEntry(PerformanceEventTiming::Candidate& candidate, Seconds duration)
+void Performance::processEventEntry(PerformanceEventTiming::Candidate& candidate)
 {
     // FIXME: handle firstInput based on InteractionId
```
Looks like this FIXME is fixed.

rreno commented on Source/WebCore/page/PerformanceEventTiming.cpp:
```diff
@@ -78,7 +79,7 @@ ASCIILiteral PerformanceEventTiming::entryType() const
 unsigned PerformanceEventTiming::interactionId() const
 {
-    return 0;
+    return m_interactionID;
```
Isn't unsigned 32 bits? m_interactionID is 64 bits.

rreno commented on Source/WebCore/page/PerformanceEventTiming.h:
```diff
@@ -39,15 +39,18 @@ class Node;
 class PerformanceEventTiming final : public PerformanceEntry {
 public:
+    using InteractionID = uint64_t;
     struct Candidate {
```
consider running `dump-class-layout` on this to make sure you've got an efficiently laid out struct.

rreno commented on Source/WebCore/page/LocalDOMWindow.cpp:
```diff
@@ -431,6 +433,8 @@ LocalDOMWindow::LocalDOMWindow(Document& document)
     ASSERT(frame());
     addLanguageChangeObserver(this, &languagesChangedCallback);
+    // User interaction value should be initialized with a random number from 100 to 10000:
+    m_userInteractionValue = 100 + WTF::cryptographicallyRandomNumber<uint64_t>() % 9901;
```
Should we only initialize this on first use? Like an `ensureUserInteractionValue` getter? I don't have a good sense for how expensive it would be to generate a cryptographically random number for every DOM window even if the event timing API is unused.

rniwa commented on Source/WebCore/page/LocalDOMWindow.h:
```diff
@@ -38,6 +38,7 @@
 #include <JavaScriptCore/HandleForward.h>
```
We shouldn't be including `PerformanceEventTiming.h` here.
We should move `InteractionID` out of `PerformanceEventTiming` and forward declare it here instead.

francovs replied:
`PerformanceEventTiming.h` is also required due `PerformanceEventTiming::Candidate`, which is used in `m_performanceEventTimingCandidates` (and others), so it cannot be forward-declared.

I believe the `PerformanceEventTiming.h` is lightweight now (it does not bring in `Performance.h` anymore) so it shouldn't be a disaster.

rniwa replied:
We should move `PerformanceEventTiming::Candidate` out of `PerformanceEventTiming` as well so that we don't need to include the header file. Generally speaking, we shouldn't be including any header files unless absolutely necessary.

francovs replied:
Moved to a separate header (`page/PerformanceEventTimingCandidate.h`)

rniwa commented on Source/WebCore/page/LocalDOMWindow.h:
```diff
@@ -440,6 +445,16 @@ class LocalDOMWindow final
+    std::optional<PerformanceEventTiming::Candidate> m_pendingPointerDown;
+
+    HashMap<int, PerformanceEventTiming::Candidate> m_pendingKeyDowns;
+    PerformanceEventTiming::InteractionID m_userInteractionValue = 99;
```
I don't think we should be using 99 as a special sentinel value. We should use `std::optional<PerformanceEventTiming::InteractionID>` instead to indicate whether interaction ID has been initialized or not.

rniwa commented on Source/WebCore/page/LocalDOMWindow.cpp:
```diff
@@ -2458,39 +2460,175 @@ void LocalDOMWindow::finishedLoading()
+    PerformanceEventTiming::InteractionID interactionId = 0;
     LOG_WITH_STREAM(PerformanceTimeline, stream << "Initializing event timing entry (type=" << event.type() << ") at t=" << processingStart);
+    // Stateful interactionId logic as described in https://www.w3.org/TR/event-timing/#sec-computing-interactionid :
+    switch (typeInfo.type()) {
```
We should extract this switch into a separate helper function named "computeInteractionId" instead of adding a comment like this.

rniwa commented on Source/WebCore/page/LocalDOMWindow.cpp:
```diff
@@ -2458,39 +2460,175 @@ void LocalDOMWindow::finishedLoading()
+
+            increaseInteractionCount();
+            auto it = m_pendingKeyDowns.find(code);
+            it->value.interactionID = ensureUserInteractionValue();
```
I would have called this function `generateUserInteractionValue` after merging it with `increaseInteractionCount`. Having separate functions to increase the count and return the current value is not a common pattern we use in WebCore.

francovs replied:
It is hard to merge these functions cleanly because `contextmenu` and `pointerup` sometimes use the current interaction count without incrementing it.

I added a return to `increaseInteractionCount()`, simplifying some logic, but kept `ensureUserInteractionValue()`. I also created `maintainInteractionCount()` to document cases in which the spec uses the current interaction value without incrementing.

rniwa commented on Source/WebCore/page/LocalDOMWindow.cpp:
```diff
@@ -2458,39 +2460,175 @@ void LocalDOMWindow::finishedLoading()
+            auto it = m_pendingKeyDowns.find(code);
+            it->value.interactionID = ensureUserInteractionValue();
+            m_performanceEventTimingCandidates.append(it->value);
+            m_pendingKeyDowns.remove(code);
```
This will do 3 hash map lookups. That's awfully inefficient.
We should be doing exactly one hash map lookup using `find` and using the resultant iterator to the rest of operations.

rniwa commented on Source/WebCore/page/LocalDOMWindow.cpp:
```diff
@@ -2458,39 +2460,175 @@ void LocalDOMWindow::finishedLoading()
+        {
+            auto inputEvent = downcast<InputEvent>(&event);
+            ASSERT(inputEvent);
+            if (!inputEvent->isInputMethodComposing())
```
So we only generate an interaction id for an input element resulted from composition??? That's very odd.

francovs replied:
I believe the idea is not to double-count an input event generated from a keydown

rniwa commented on Source/WebCore/page/LocalDOMWindow.cpp:
```diff
@@ -2458,39 +2460,175 @@ void LocalDOMWindow::finishedLoading()
+        if (!m_pointerInteractionId)
+            break;
+        interactionId = *m_pointerInteractionId;
+        m_pointerInteractionId.reset();
```
Where is `pointerMap[pointerId]` check that's in the spec?

francovs replied:
I renamed `pointerMap` to `pointerInteractionId` since it is not a map, as we only support a single pointer. `pointerMap` sounds very bad to me considering `pendingPointerDowns` is also supposed to be a map.

I think it would be better to merge `m_pendingPointerDown` and `m_pointerInteractionId` into a single struct containing both (perhaps `PointerInteractionState`), and make them into a single map when multiple pointer support is added. That would also save a few lookups.

francovs replied:
Implemented as follows:

```C++
struct PointerInteractionState {
    std::optional<InteractionID> interactionID;
    std::optional<PerformanceEventTimingCandidate> pendingPointerDown;
};
```

rniwa commented on Source/WebCore/page/PerformanceEventTimingCandidate.h:
```diff
@@ -0,0 +1,47 @@
+
+namespace WebCore {
+
+using InteractionID = uint64_t;
```
We should probably wrap this in `struct` so that the type check kicks in.
It doesn't seem right that this uses `uint64_t`. The number in JS doesn't have 64-bit precision.
We should instead limit the value at UINT_MAX.

francovs replied:
Strong typing added! However, limiting to 32-bit sounds wrong to me because the IDL spec [specifically requests 64-bit unsigned](https://www.w3.org/TR/event-timing/#sec-extensions). IMO this means the JS Wrapper should be producing a javascript BigInt instead.

rniwa commented on Source/WebCore/page/LocalDOMWindow.h:
```diff
@@ -396,6 +398,11 @@ class LocalDOMWindow final
 #endif
 
+    InteractionID computeInteractionID(const Event&, EventType);
+    InteractionID& ensureUserInteractionValue();
```
This should be a private function now that it's only called by other member functions.

francovs replied:
These are all private already; only `interactionCount()` is currently public.

rniwa commented on Source/WebCore/page/LocalDOMWindow.cpp:
```diff
@@ -2458,39 +2460,188 @@ void LocalDOMWindow::finishedLoading()
+    return m_userInteractionValue.value();
+}
+
+InteractionID LocalDOMWindow::increaseInteractionCount()
```
I'd call this `generateInteractionId`

rniwa commented on Source/WebCore/page/LocalDOMWindow.cpp:
```diff
@@ -2458,39 +2460,188 @@ void LocalDOMWindow::finishedLoading()
+    return ensureUserInteractionValue();
+}
+
+InteractionID LocalDOMWindow::maintainInteractionCount()
```
We should probably call this `currentInteractionId`.

rniwa commented on Source/WebCore/page/LocalDOMWindow.cpp:
```diff
@@ -2458,39 +2460,188 @@ void LocalDOMWindow::finishedLoading()
+        if (it == m_pendingKeyDowns.end())
+            return 0;
+
+        auto ret = increaseInteractionCount();
```
Please don't use abbreviations like `ret`. Spell out `interactionId`.

francovs replied:
Done! (removed `temp` as well)

rniwa commented on Source/WebCore/page/LocalDOMWindow.cpp:
```diff
@@ -2458,39 +2460,188 @@ void LocalDOMWindow::finishedLoading()
+        return increaseInteractionCount();
+    }
+    case EventType::click: {
+        if (!m_pointerInteractionState.interactionID)
```
Why do we not have `pointerMap`? We should at least clarify why in the commit message.

francovs replied:
Commit message updated with the following:

```
We currently only support a single pointer, so the map variables
related to pointer interactions (pointerMap[] and pendingPointerDown[])
mentioned in the spec (https://www.w3.org/TR/event-timing/#pointer-interaction-value-map)
were implemented as a struct (PointerInteractionState) containing
the necessary state for a single pointer. In that conversion,
pointerMap was renamed to PointerInteractionState::interactionID.
```

rniwa commented on Source/WebCore/page/PerformanceEventTimingCandidate.h:
```diff
@@ -0,0 +1,51 @@
+namespace WebCore {
+
+struct InteractionID {
+    uint64_t value = 0;
```
Use `{ 0 }` for consistency?

rniwa commented on Source/WebCore/page/PerformanceEventTiming.h:
```diff
@@ -26,35 +26,29 @@
+#include "EventNames.h"
+#include "EventTarget.h"
+#include "PerformanceEntry.h"
+#include "PerformanceEventTimingCandidate.h"
```
I don't think we need to include this header. Just upward declare `PerformanceEventTimingCandidate`.

francovs replied:
Done by splitting `InteractionID` in its own header. I took the opportunity to rename it to `EventTimingInteractionID`.

rniwa commented on Source/WebCore/page/PerformanceEventTimingCandidate.h:
```diff
@@ -0,0 +1,51 @@
+
+#pragma once
+
+#include "EventNames.h"
```
We shouldn't have to include `EventNames.h`.

rniwa commented on Source/WebCore/page/LocalDOMWindow.cpp:
```diff
@@ -2458,39 +2460,189 @@ void LocalDOMWindow::finishedLoading()
+            return currentInteractionID();
+
+        m_contextMenuTriggered = true;
+        auto interactionID = generateInteractionID();
```
This sequence of code is duplicated for `EventType::pointerup` and `EventType::contextmenu`.
Can we extract a helper function / lambda so that the code is shared?

rniwa commented on Source/WebCore/page/LocalDOMWindow.h:
```diff
@@ -438,7 +444,21 @@ class LocalDOMWindow final
+    Vector<PerformanceEventTimingCandidate, 6> m_performanceEventTimingCandidates;
+
+    struct PointerInteractionState {
+        EventTimingInteractionID interactionID;
```
I think we should call this pointerMapEquivalent or something so that it's easily recognizable as a replacement for pointerMap in the spec. I also think we should use Markable<EventTimingInteractionID> or std::optional< EventTimingInteractionID> instead of using the 0 as the sentinel value.

francovs replied:
I rewrote this to better mirror the spec:

```C++
// FIXME: support multiple pointers by replacing std::optional
// with map objects using PointerID as key:
std::optional<EventTimingInteractionID> m_pointerMap;
std::optional<PerformanceEventTimingCandidate> m_pendingPointerDown;
```

The idea of a separate struct was to try to better accommodate multiple pointers with something like `HashMap<PointerID, PointerInteractionState> pointerMap`. It doesn't really work well with the current code.

rniwa replied:
That sounds fine. We probably want to use `Markable` instead for space efficiency though. We can use 0 as the empty value.

rniwa commented on Source/WebCore/page/LocalDOMWindow.h:
```diff
@@ -438,7 +444,21 @@ class LocalDOMWindow final
-    Vector<PerformanceEventTiming::Candidate, 6> m_performanceEventTimingCandidates;
+    Vector<PerformanceEventTimingCandidate, 6> m_performanceEventTimingCandidates;
+
+    struct PointerInteractionState {
```
Oh, I didn't realize we're using a nested struct here.
Why do need this stunt? It doesn't seem like it's making the code any more clearer.
I suggest we store two values directly in the outer class.

rniwa commented on Source/WebCore/page/LocalDOMWindow.h:
```diff
@@ -438,7 +444,21 @@ class LocalDOMWindow final
+    // m_pointerInteractionState as a map with PointerID as key:
+    PointerInteractionState m_pointerInteractionState;
+
+    bool m_contextMenuTriggered = false;
```
Use `{ ~ }` instead.

rniwa commented on Source/WebCore/page/LocalDOMWindow.h:
```diff
@@ -438,7 +444,21 @@ class LocalDOMWindow final
+
+    HashMap<int, PerformanceEventTimingCandidate> m_pendingKeyDowns;
+    EventTimingInteractionID m_userInteractionValue;
+    uint64_t m_interactionCount = 0;
```
Ditto.

rniwa commented on Source/WebCore/page/LocalDOMWindow.cpp:
```diff
@@ -2458,39 +2461,193 @@ void LocalDOMWindow::finishedLoading()
+{
+    // User interaction value should be increased by a small integer to
+    // "discourage developers from considering it as a counter":
+    ensureUserInteractionValue().value += 7;
```
Hm... doesn't this mean the developer could just divide the number by 7 to get an effective counter?
Do other browsers hard-code a value like this?

francovs replied:
Yes, though this isn't really a problem since `window.interactionCount` is available. The spec explicitly allows a constant increment, it only warns agains exposing leaks:

> A user agent may choose to increase it by a small random integer every time, or choose a constant. A user agent must not use a shared global [user interaction value](https://www.w3.org/TR/event-timing/#user-interaction-value)s for all [Windows](https://html.spec.whatwg.org/multipage/nav-history-apis.html#window), because this could introduce cross-origin leaks.

My understanding is that the point is to provide a soft warning to a user trying to match `PerformanceEventTiming.interactionId` to `window.interactionCount`. For that, maybe it would be better to pick something around 100, which would ensure these two are at least two orders of magnitude apart when considering the latest interaction.

I picked 7 on a whim because it is small and has no obvious divisibility test, but it turns out both [chromium](https://source.chromium.org/chromium/chromium/src/+/main:third_party/blink/renderer/core/timing/responsiveness_metrics.cc;drc=534e2701fc2faaa3057ab8bf882cfc44ff6a4b9b;l=274) and [Gecko](https://github.com/mozilla/gecko-dev/blob/5836a062726f715fda621338a17b51aff30d0a8c/dom/performance/PerformanceInteractionMetrics.cpp#L14C1-L18C48) picked the same number:

```C++
// Interaction ID increment. We increase this value by an integer greater than 1
// to discourage developers from using the value to 'count' the number of user
// interactions. This is consistent with the spec, which allows the increasing
// the user interaction value by a small number chosen by the user agent.
constexpr uint32_t kInteractionIdIncrement = 7;
```

rniwa commented on Source/WebCore/page/LocalDOMWindow.h:
```diff
@@ -438,7 +445,19 @@ class LocalDOMWindow final
+    bool m_contextMenuTriggered { false };
+
+    // StdMap is used instead of HashMap to avoid collisions with the empty key:
+    StdMap<int, PerformanceEventTimingCandidate> m_pendingKeyDowns;
```
We should use just use `SignedWithZeroKeyHashTraits`. See `PointerCaptureController` for an example.

rniwa commented on Source/WebCore/page/Performance.cpp:
```diff
@@ -534,8 +557,9 @@ void Performance::queueEntry(PerformanceEntry& entry)
 
     if (!shouldScheduleTask)
-        return;
+    return;
```
This looks wrong.

rniwa commented on Source/WebCore/page/PerformanceObserver.h:
```diff
@@ -83,10 +83,10 @@ class PerformanceObserver : public RefCounted<PerformanceObserver> {
     Vector<Ref<PerformanceEntry>> m_entriesToDeliver;
     const Ref<PerformanceObserverCallback> m_callback;
     OptionSet<PerformanceEntry::Type> m_typeFilter;
+    Seconds m_durationThreshold { 0 };
```
No need to initialize this explicitly to 0. The default constructor does that automatically.

rniwa commented on Source/WebCore/page/PerformanceEventTimingCandidate.h:
```diff
@@ -0,0 +1,47 @@
+struct PerformanceEventTimingCandidate {
+    EventType type { };
+    bool cancelable { false };
+    Seconds startTime { 0 };
```
I don't think you need to explicitly initialize `Seconds` like this

rniwa replied:
This comment will withstands.

francovs replied:
Now fixed, but removing the explicit initializer caused [some bots](https://ews-build.webkit.org/#/builders/5/builds/123576) to hit `-Werror=missing-field-initializers`. I had to add some unnecessary stuff to the designated constructor call to get them to build:

```C++
// LocalDOMWindow::initializeEventTimingEntry():
    return PerformanceEventTimingCandidate {
        .type = type,
        .cancelable = event.cancelable(),
        .startTime =  startTime,
        .processingStart = processingStart,
        .processingEnd = { }, // here
        .duration = { }, // here
        .target = { }, // here
        .interactionID = computeInteractionID(event, type)
    };
```

I could also add explicit constructors to `PerformanceEventTimingCandidate` to avoid all this, but it felt very boilerplate-y.

--

I think this is a bad warning since missing elements [should be copy initialized from `{ }`](https://en.cppreference.com/w/cpp/language/aggregate_initialization.html), which feels safe to me (see "Implicitly initialized elements"). [This example in Godbolt](https://godbolt.org/#g:!((g:!((g:!((h:codeEditor,i:(filename:'1',fontScale:14,fontUsePx:'0',j:1,lang:c%2B%2B,selection:(endColumn:29,endLineNumber:23,positionColumn:29,positionLineNumber:23,selectionStartColumn:29,selectionStartLineNumber:23,startColumn:29,startLineNumber:23),source:'%0A//+To+reproduce+the+warning,+compile+on+clang%0A//+with+--std%3Dc%2B%2B23+-Wmissing-field-initializers%0A%0Aclass+OpaqueNumber%0A%7B%0Apublic:%0A++++int+get()+%7B+return+inner_value%3B+%7D%0Aprivate:%0A++++int+inner_value+%7B+111+%7D%3B%0A%7D%3B%0A%0Astruct+DesignatedInitializerStruct%0A%7B%0A++++int+foo%3B%0A++++//+bar+does+not+have+a+member+initializer+and+is+not+a+reference,+so%0A++++//+it+should+be+copy+initialized+from+an+empty+initializer+list.%0A++++//%0A++++//+See+https://en.cppreference.com/w/cpp/language/aggregate_initialization.html%0A++++//+(%22Implicitly+initialized+elements%22+section)%0A++++OpaqueNumber+bar%3B%0A++++//+Using+this+one+stops+the+warning:%0A++++//+OpaqueNumber+bar+%7B+%7D%3B%0A%7D%3B%0A%0Aint+main()%0A%7B%0A++++//+This+should+not+produce+a+warning:%0A++++DesignatedInitializerStruct+test+%7B+.foo+%3D+222+%7D%3B%0A++++return+test.bar.get()%3B%0A%7D%0A'),l:'5',n:'0',o:'C%2B%2B+source+%231',t:'0')),k:52.176541717049574,l:'4',n:'0',o:'',s:0,t:'0'),(g:!((g:!((h:output,i:(compilerName:'x86-64+gcc+15.1',editorid:1,fontScale:14,fontUsePx:'0',j:1,wrap:'1'),l:'5',n:'0',o:'Output+of+armv8-a+clang+(all+architectural+features,+trunk)+(Compiler+%231)',t:'0')),header:(),l:'4',m:50,n:'0',o:'',s:0,t:'0'),(g:!((h:compiler,i:(compiler:armv8-full-clang-trunk,filters:(b:'0',binary:'1',binaryObject:'1',commentOnly:'0',debugCalls:'1',demangle:'0',directives:'0',execute:'0',intel:'0',libraryCode:'0',trim:'1',verboseDemangling:'0'),flagsViewOpen:'1',fontScale:14,fontUsePx:'0',j:1,lang:c%2B%2B,libs:!(),options:'--std%3Dc%2B%2B23+-Wmissing-field-initializers',overrides:!(),selection:(endColumn:1,endLineNumber:1,positionColumn:1,positionLineNumber:1,selectionStartColumn:1,selectionStartLineNumber:1,startColumn:1,startLineNumber:1),source:1),l:'5',n:'0',o:'+armv8-a+clang+(all+architectural+features,+trunk)+(Editor+%231)',t:'0')),k:53.18242876041622,l:'4',m:50,n:'0',o:'',s:0,t:'0')),k:47.823458282950426,l:'3',n:'0',o:'',t:'0')),l:'2',n:'0',o:'',t:'0')),version:4) reproduces the warning.

rniwa commented on Source/WebCore/page/LocalDOMWindow.h:
```diff
@@ -30,16 +30,16 @@
 #include <wtf/Function.h>
 #include <wtf/HashSet.h>
 #include <wtf/MonotonicTime.h>
+#include <wtf/StdMap.h>
```
No need to include `StdMap.h` now.

francovs committed:
Event Timing: implement interactionId logic
https://bugs.webkit.org/show_bug.cgi?id=297361
rdar://158264903

Reviewed by Ryosuke Niwa.

Assigns interactionId to PerformanceEventTiming entries that qualify,
according to https://www.w3.org/TR/event-timing/.

Only certain deliberate user actions such as clicks or keypresses are
considered; passive events such as pointerover or pointerexit do not
qualify. Some events are grouped together, receiving the same
interactionId. One such example is pointerdown, pointerup and click.

We currently only support a single pointer, so the map variables
related to pointer interactions (pointerMap[] and pendingPointerDown[])
mentioned in the spec (https://www.w3.org/TR/event-timing/#pointer-interaction-value-map)
were implemented as a struct (PointerInteractionState) containing
the necessary state for a single pointer. In that conversion,
pointerMap was renamed to PointerInteractionState::interactionID.

New files created to achieve more fine grained header inclusion:
* Source/WebCore/page/EventTimingInteractionID.h
* Source/WebCore/page/PerformanceEventTimingCandidate.h

Canonical link: https://commits.webkit.org/299033@main
